File: codes/PancakeSort969/Solution.java
package codes.PancakeSort969;
import java.util.ArrayList;
import java.util.List;

class Solution {
    //煎饼排序，注意的是数组内元素最大值的是数组的长度，根据最大值递减找元素来排序
    public List<Integer> pancakeSort(int[] arr) { 
        List<Integer> ans = new ArrayList<Integer>();
        int arrLen = arr.length;
        int last = arrLen;
        for(int i = arrLen - 1; i >= 0 ; i--){
            if(arr[i] == last){
                last--;
                
                continue;
            }
            if(arr[0] == last){
                reverse(arr, 0, i);
                last--;
                ans.add(i+1);
                continue;
            }
            for(int k = 0; k < last; k++){
                if(arr[k] != last){
                    continue;
                }
                reverse(arr, 0, k);
                ans.add(k+1);
                reverse(arr, 0, i);
                ans.add(i+1);
                break;
            }
            last--;
            
        }
        
        return ans;
    }

    public void reverse(int[] arr, int startIndex, int endIndex){
        for(int i = startIndex; i<(endIndex - startIndex+1)/2; i++){
            int temp = arr[i];
            arr[i] = arr[endIndex-i];
            arr[endIndex-i] = temp;
        }
    }

    public void swap(int num1, int num2){
        int temp = num1;
        num1 = num2;
        num2 = temp;
    }

    public static void main(String[] args) {
        int[] nums = {3,2,4,1};
        Solution solution = new Solution();
        System.out.println(solution.pancakeSort(nums));
    }
}

File: codes/PancakeSort969/solution.cs
public class Solution {
    public IList<int> PancakeSort(int[] arr) {
        IList<int> res = new List<int>();
        for(int n = arr.Length; n > 1; -- n){
            int idx = 0;
            // 寻找当前范围最大值
            for(int i = 1; i < n; ++ i){
                if(arr[i] >= arr[idx]){
                    idx = i;
                }
            }
            if(idx == n - 1){
                // 最大值已在末尾处，无需翻转
                continue;
            }
            // 两步翻转将最大值移至末尾
            Reverse(arr, idx);
            Reverse(arr, n - 1);
            // 记录k值序列
            res.Add(idx + 1);
            res.Add(n);
        }
        return res;
    }

    public void Reverse(int[] arr, int range){
        for(int i = 0, j = range; i < j; ++ i, -- j){
            int temp = arr[i];
            arr[i] = arr[j];
            arr[j] = temp;
        }
    }
}